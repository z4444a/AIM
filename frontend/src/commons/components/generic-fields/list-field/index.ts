import { ListField } from './list-field';

export default ListField;
