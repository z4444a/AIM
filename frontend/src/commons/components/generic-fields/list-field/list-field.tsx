import React, { ReactNode } from 'react';
import TextField from '@material-ui/core/TextField';
import MenuItem from '@material-ui/core/MenuItem';
import {
  ListFieldConfigModel,
  ListFieldHandlers,
  ListFieldValue,
} from '../models/field-config.model';
import AimMultipleSelect from '../../aim-multiple-select/index';
import NamedModel from '../../../../model/base/named-model';
import { ItemSelection } from '../../aim-multiple-select/aim-multiple-select';
import _ from 'underscore';

export interface Props {
  fieldConfig: ListFieldConfigModel;
  handlers: ListFieldHandlers;
  value?: ListFieldValue;
  values?: ListFieldValue[];
}

export type InternalProps = Props;

export class ListField extends React.PureComponent<InternalProps, ItemSelection> {
  private readonly emptyValue = '';

  constructor(props: InternalProps) {
    super(props);
    this.state = ListField.getSelection(
      props.fieldConfig.optionList,
      props.values ? props.values.map(item => item.toString()) : []
    );
  }

  componentDidUpdate(prevProps: Readonly<InternalProps>, prevState: Readonly<ItemSelection>): void {
    const { values, fieldConfig } = this.props;
    if (
      !_.isEqual(values, prevProps.values) ||
      !_.isEqual(fieldConfig.optionList, prevProps.fieldConfig.optionList)
    ) {
      this.setState(
        ListField.getSelection(
          fieldConfig.optionList,
          values ? values.map(item => item.toString()) : []
        )
      );
    }
  }

  public render(): ReactNode {
    const { fieldConfig, value } = this.props;
    const {
      className,
      style,
      required,
      key,
      optionList,
      labelConfig,
      error,
      multipleMax,
    } = fieldConfig;
    const multiple = multipleMax && multipleMax > 1;
    const { available, selected } = this.state;
    return multiple ? (
      <AimMultipleSelect
        disabled={selected.length === multipleMax || available.length === 0}
        available={available}
        selected={selected}
        selectItems={this.selectItems}
        label={labelConfig && labelConfig.visible ? labelConfig.value : ''}
        error={!!error}
      />
    ) : (
      <TextField
        select
        style={style}
        className={className}
        required={required}
        error={!!error}
        helperText={error ? error.value : undefined}
        value={value ? value : this.emptyValue}
        onChange={this.handleValueChange}
        label={labelConfig && labelConfig.visible ? labelConfig.value : undefined}
      >
        {Object.keys(optionList).map(itemKey => (
          <MenuItem key={`${key}_${itemKey}`} value={itemKey}>
            {optionList[itemKey]}
          </MenuItem>
        ))}
      </TextField>
    );
  }

  private static getSelection = (
    options: { [key: string]: string },
    values: ListFieldValue[]
  ): ItemSelection => {
    const indexList = Object.keys(options);
    return {
      available: indexList
        .filter(item => values.indexOf(item) < 0)
        .map(item => ListField.toNamed(options, item)),
      selected: indexList
        .filter(item => values.indexOf(item) >= 0)
        .map(item => ListField.toNamed(options, item)),
    };
  };

  private static toNamed = (options: { [key: string]: string }, value: string): NamedModel => {
    return {
      id: parseInt(value),
      name: options[value],
    };
  };

  private selectItems = (selection: ItemSelection) => {
    this.setState(selection);
    const { handlers, fieldConfig } = this.props;
    if (!handlers || typeof handlers.onValuesChange !== 'function') {
      return;
    }
    handlers.onValuesChange(selection.selected.map(item => item.id.toString()), fieldConfig.key);
  };

  private handleValueChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { handlers, fieldConfig } = this.props;
    if (!handlers || typeof handlers.onValueChange !== 'function') {
      return;
    }
    handlers.onValueChange(event.target.value, fieldConfig.key);
  };
}
