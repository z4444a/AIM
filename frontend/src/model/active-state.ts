export const ActiveState = [
  {
    key: '1',
    value: 'Активен',
  },
  {
    key: '2',
    value: 'Не активен',
  },
];
