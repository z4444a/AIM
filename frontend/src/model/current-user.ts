import { Role } from '../commons/role';

export interface CurrentUser {
  id: number;
  fullName: string;
  username: string;
  role: Role;
}
