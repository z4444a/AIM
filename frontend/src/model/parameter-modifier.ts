export enum ParameterModifier {
  POOL_PARAMETER = 1,
  REQUEST_PARAMETER = 2,
  ALLOCATION_PARAMETER = 3,
}
