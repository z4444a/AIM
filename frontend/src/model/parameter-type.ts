export enum ParameterType {
  NUMBER = 'NUMBER',
  TEXT = 'TEXT',
  DATE = 'DATE',
  LIST = 'LIST',
  REAL = 'REAL',
  POOL = 'POOL',
}
