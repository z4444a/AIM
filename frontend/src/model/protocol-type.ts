export const Protocols = [
  {
    key: '1',
    value: 'TLS',
  },
  {
    key: '2',
    value: 'SSL',
  },
  {
    key: '3',
    value: 'None',
  },
];
