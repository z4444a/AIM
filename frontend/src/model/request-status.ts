import { RequestStatus } from './get/request-status';

export function RequestStatuses(t: (s: string) => string) {
  return [
    {
      key: RequestStatus.NEW.toString(),
      value: t('requestPage.status.new'),
    },
    {
      key: RequestStatus.IN_PROGRESS.toString(),
      value: t('requestPage.status.inWork'),
    },
    {
      key: RequestStatus.PROCESSED.toString(),
      value: t('requestPage.status.executed'),
    },
    {
      key: RequestStatus.CANCELED.toString(),
      value: t('requestPage.status.canceled'),
    },
    {
      key: RequestStatus.PAUSED.toString(),
      value: t('requestPage.status.paused'),
    },
  ];
}
