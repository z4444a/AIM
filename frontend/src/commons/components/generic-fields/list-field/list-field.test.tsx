import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { ListFieldConfigModel, ListFieldHandlers } from '../models/field-config.model';
import { InternalProps, ListField } from './list-field';

describe('ListField', () => {
  let wrapper: ShallowWrapper;
  beforeEach(() => {
    const fieldConfig = new ListFieldConfigModel('listField');
    fieldConfig.optionList = {
      firstOption: 'FirstOption',
      secondOption: 'SecondOption',
    };
    const handlers: ListFieldHandlers = {
      onValueChange: () => {},
    };

    const props: InternalProps = {
      fieldConfig,
      handlers,
    };
    wrapper = shallow(<ListField {...props} />);
  });

  it('should render correctly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
