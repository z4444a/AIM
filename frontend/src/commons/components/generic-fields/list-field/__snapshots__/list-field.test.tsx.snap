// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListField should render correctly 1`] = `
<TextField
  error={false}
  onChange={[Function]}
  required={false}
  select={true}
  value=""
  variant="standard"
>
  <WithStyles(MenuItem)
    key="listField_firstOption"
    value="firstOption"
  >
    FirstOption
  </WithStyles(MenuItem)>
  <WithStyles(MenuItem)
    key="listField_secondOption"
    value="secondOption"
  >
    SecondOption
  </WithStyles(MenuItem)>
</TextField>
`;
